File: router/workers/engagement.py
"""Engagement path — rolling like/repost counts, alerting on unusual traction.

This is the highest-volume path by a wide margin. Measured over 440,108
classified events: engagement 79.3% of routed traffic, against 5.6% for the
graph path — a 14.2x spread. A design that gave every path the same capacity
would be simultaneously wasteful on the graph path and starved on this one,
which is the concrete reason the paths scale independently rather than sharing
a pool.

Counting happens per *target* post, not per actor: the question is "is this post
getting unusual traction", so likes and reposts of the same URI accumulate
together.
"""

from __future__ import annotations

from typing import Any

from ..observability import get_logger
from .runner import WorkerContext

log = get_logger("engagement")

NAMESPACE = "eng"


def subject_uri(event: dict[str, Any]) -> str | None:
    """Extract the post being liked/reposted.

    Verified shape: ``commit.record.subject = {"uri": "at://...", "cid": "..."}``
    for both ``app.bsky.feed.like`` and ``app.bsky.feed.repost``. Note this
    differs from ``app.bsky.graph.follow``, where ``subject`` is a bare DID
    string — hence a dedicated extractor per path rather than one clever shared one.
    """
    record = (event.get("commit") or {}).get("record") or {}
    subject = record.get("subject")
    if isinstance(subject, dict):
        uri = subject.get("uri")
        return uri if isinstance(uri, str) and uri else None
    return None


class EngagementHandler:
    name = "engagement"

    def __init__(self) -> None:
        self.ctx: WorkerContext | None = None

    async def setup(self, ctx: WorkerContext) -> None:
        self.ctx = ctx
        log.info(
            "engagement handler ready",
            extra={"context": {"window_s": ctx.settings.engagement_window_s,
                               "threshold": ctx.settings.engagement_threshold}},
        )

    async def handle(self, event: dict[str, Any], headers: dict[str, str]) -> None:
        assert self.ctx is not None
        uri = subject_uri(event)
        if uri is None:
            # A like with no subject is malformed for our purposes, but it is
            # not worth a retry or a DLQ entry — just not countable.
            return

        settings = self.ctx.settings
        total = await self.ctx.store.bump_and_total(
            NAMESPACE, uri, settings.engagement_window_s
        )
        if total < settings.engagement_threshold:
            return

        # Threshold crossed. Fire once per window per URI, not once per event.
        if not await self.ctx.store.claim_alert(NAMESPACE, uri, settings.engagement_window_s):
            return

        self.ctx.alert(
            self.name,
            "unusual-traction",
            uri=uri,
            count=total,
            window_s=settings.engagement_window_s,
            threshold=settings.engagement_threshold,
            last_actor=event.get("did"),
            kind=(event.get("commit") or {}).get("collection"),
        )

File: internal/handler/engagement.go
package handler

import (
	"context"

	"github.com/shalevsharfsh/jetstream-router/internal/event"
	"github.com/shalevsharfsh/jetstream-router/internal/obs"
)

// Engagement keeps rolling per-subject counts for likes and reposts and raises
// an alert when one crosses a threshold.
//
// The highest-volume route by a wide margin. Counting is per *target* post, not
// per actor: the question is whether a post is getting unusual traction, so
// likes and reposts of the same URI accumulate together.
type Engagement struct {
	sink      Sink
	threshold int
	windows   []*Window
	throttles []*Throttle
	dedup     *dedupGate
}

// NewEngagement builds the route's handler with one state shard per worker.
func NewEngagement(sink Sink, shards, threshold int, windowUS int64, l Limits) *Engagement {
	if shards < 1 {
		shards = 1
	}
	e := &Engagement{sink: sink, threshold: threshold, dedup: newDedupGate("engagement", shards, l)}
	for i := 0; i < shards; i++ {
		e.windows = append(e.windows, NewWindow(windowUS, l.MaxKeysPerShard, evictCounter("engagement")))
		e.throttles = append(e.throttles, NewThrottle(windowUS, l.MaxKeysPerShard, evictCounter("engagement")))
	}
	return e
}

// ShardKey partitions by the post being engaged with, so every count for one
// subject is owned by one goroutine.
func (e *Engagement) ShardKey(ev event.Event) string {
	var rec engagementRecord
	if !decode(ev.Record, &rec) {
		return ""
	}
	return rec.Subject.URI
}

// Handle counts one like or repost.
func (e *Engagement) Handle(_ context.Context, shard int, ev event.Event) error {
	if shard >= len(e.windows) || shard < 0 {
		shard = 0
	}
	var rec engagementRecord
	if !decode(ev.Record, &rec) || rec.Subject.URI == "" {
		// Not countable, but not worth a retry or a dead letter either.
		return nil
	}
	if e.dedup.duplicate(shard, ev) {
		return nil
	}

	n := e.windows[shard].Add(rec.Subject.URI, ev.TimeUS)
	obs.StateEntries.WithLabelValues("engagement").Set(float64(e.windows[shard].Len()))

	if n < e.threshold {
		return nil
	}
	// Thresholds can be crossed deliberately, so an unthrottled alert path is a
	// remotely triggerable flood against whatever sits downstream of it.
	if !e.throttles[shard].Allow(rec.Subject.URI, ev.TimeUS) {
		obs.AlertsThrottled.WithLabelValues("engagement").Inc()
		return nil
	}

	e.sink.Alert("warn", "engagement", "threshold crossed",
		"subject", rec.Subject.URI, "count", n,
		"threshold", e.threshold, "collection", ev.Key.Collection)
	return nil
}
